### PR TITLE
Add service to swap two parent structures.

### PR DESCRIPTION
**Pull request by iamanandkr**
Create the service to swap two parent structures. Swapping of structures is only allowed if the two parent structures are unique or just duplicates of each other than the swapping.

## Description

Recommended workflow to check:
1. ParentSwapStructuresDTO > ParentSwapStructures
2.  APIParentController > swapParentStructures
3. ParentSwapStructureServiceImpl > swapParentStructurs.

## Related Issue

[ACAS-286](https://jira.schrodinger.com/browse/ACAS-286).

## How Has This Been Tested?

See testing section in the acas PR.

## TODO
~I am seeing issues when querying duplicate parents i.e in the cases where duplicate parents do exist(confirmed via the `/validateParent` API). I am not getting the duplicate parents listed in the `swapParentStructures` service.~

### CODE REVIEW

File: src/main/java/com/labsynch/labseer/api/ApiParentController.java [hunk 1]
```diff
@@ -67,6 +74,31 @@ public ResponseEntity<String> updateParent(@RequestBody String json) {
...
+			"corpName1=%s & corpName2=%s swap: ",
+			parentSwapStructuresDTO.getCorpName1(),
+			parentSwapStructuresDTO.getCorpName2());
+		try {
+			if (parentSwapStructuresService.swapParentStructures(parentSwapStructuresDTO)) {
+				logger.info(msg + "success");
+				return new ResponseEntity<String>(headers, HttpStatus.OK);
+			} else {
+				logger.info(msg + "failure");
+				return new ResponseEntity<String>(headers, HttpStatus.INTERNAL_SERVER_ERROR);
```
Comment (dalejerikson) [inline src/main/java/com/labsynch/labseer/api/ApiParentController.java:94]: Everything looks great and clean so far! I would suggest adding a JSON message in the `ResponseEntity` for both a successful swap or (even better) an error

File: src/main/java/com/labsynch/labseer/service/ParentSwapStructuresServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+    Logger logger = LoggerFactory.getLogger(ParentSwapStructuresServiceImpl.class);
+
+    @Autowired
+    public ParentService parentService;
+
+    @Override
+    public boolean swapParentStructures(ParentSwapStructuresDTO parentSwapStructuresDTO) {
+        String corpName1 = parentSwapStructuresDTO.getCorpName1();
+		String corpName2 = parentSwapStructuresDTO.getCorpName2();
```
Comment (dalejerikson) [inline src/main/java/com/labsynch/labseer/service/ParentSwapStructuresServiceImpl.java:25]: Fix Indentation

File: src/main/java/com/labsynch/labseer/service/ParentSwapStructuresServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+		} catch (Exception e) {
+			logger.error("Caught error while fetching parent", e);
+            return false;
+		}
+
+		ParentValidationDTO validationDTO1;
+		ParentValidationDTO validationDTO2;
+		try {
+			validationDTO1 = parentService.validateUniqueParent(parent1);
+			validationDTO2 = parentService.validateUniqueParent(parent2);
```
Comment (dalejerikson) [inline src/main/java/com/labsynch/labseer/service/ParentSwapStructuresServiceImpl.java:41]: I lack knowledge of this but is `validateUniqueParent` needed if it was instantiated by `findParentsByCorpNameEquals` above

**Comment by dalejerikson** [comment]
Nice clean code! Easy to read and self explanatory. Just have one question and a few minor suggested changes.

**Review by dalejerikson — DISMISSED** [review]

**Comment by iamanandkr** [comment]
Please hold off reviewing the PR. I need to fix things in here.

**Comment by iamanandkr** [comment]
> Please hold off reviewing the PR. I need to fix things in here.

PR is good to be reviewed for the code logic.

File: src/main/java/com/labsynch/labseer/service/ParentSwapStructuresServiceImpl.java (hunk 3 above)
Comment (iamanandkr) [reply src/main/java/com/labsynch/labseer/service/ParentSwapStructuresServiceImpl.java:48]: If the item exists in the db then it should already be valid and here the logic is merely sanity check. I am open to removing the check from here entirely.

File: src/main/java/com/labsynch/labseer/service/ParentSwapStructuresServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+		for (int parentCdId : parentCdIds) {
+			for (Parent parent : Parent.findParentsByCdId(parentCdId).getResultList()) {
+				String parentStereoCategory = parent.getStereoCategory().getCode();
+				if (!parentStereoCategory.equalsIgnoreCase(stereoCategory)) {
+					continue;
+				}
+				String parentStereoComment = parent.getStereoComment();
+				if (parentStereoComment == null && stereoComment == null) {
+					matchingParents.add(parent);
+				} else if (parentStereoComment == null || stereoComment == null) {
```
Comment (bffrost) [inline src/main/java/com/labsynch/labseer/service/ParentSwapStructuresServiceImpl.java:104]: I think there's a possible complication when one is null and the other is empty string. Users don't have a way to truly null out a stereo comment, so we should consider null and empty string as equal to each other.
We handle this case elsewhere, although not that elegantly: https://github.com/mcneilco/acas-roo-server/blob/release/2022.1.x/src/main/java/com/labsynch/labseer/service/BulkLoadServiceImpl.java#L806

**Review by bffrost — CHANGES_REQUESTED** [review]
Spotted a small unhandled edge case. Otherwise looks good to me!

File: src/main/java/com/labsynch/labseer/api/ApiParentController.java (hunk 1 above)
Comment (iamanandkr) [reply src/main/java/com/labsynch/labseer/api/ApiParentController.java]: Done. Let me know if the output response looks good or should I add/remove any fields. cc'ing @bffrost too!

File: src/main/java/com/labsynch/labseer/service/ParentSwapStructuresServiceImpl.java (hunk 3 above)
Comment (iamanandkr) [reply src/main/java/com/labsynch/labseer/service/ParentSwapStructuresServiceImpl.java:50]: I am keeping the checks for now!

**Pull request merged**